File: apps/macos/Mollotov/Handlers/NetworkInspectorHandler.swift
import Foundation

/// API handler for network inspector: list/filter traffic, get detail, navigate to entry, clear.
struct NetworkInspectorHandler {
    let context: HandlerContext

    func register(on router: Router) {
        router.register("network-list") { body in await list(body) }
        router.register("network-detail") { body in await detail(body) }
        router.register("network-select") { body in await select(body) }
        router.register("network-current") { _ in await current() }
        router.register("network-clear") { _ in await clear() }
    }

    @MainActor
    private func list(_ body: [String: Any]) async -> [String: Any] {
        let store = NetworkTrafficStore.shared
        let entries = store.toSummaryJSON(
            method: body["method"] as? String,
            category: body["category"] as? String,
            statusRange: body["statusRange"] as? String,
            urlPattern: body["urlPattern"] as? String
        )
        return successResponse(["entries": entries, "total": store.entries.count])
    }

    @MainActor
    private func detail(_ body: [String: Any]) async -> [String: Any] {
        let store = NetworkTrafficStore.shared
        guard let index = body["index"] as? Int, store.entries.indices.contains(index) else {
            return errorResponse(code: "INVALID_INDEX", message: "index is required and must be valid")
        }
        return successResponse(store.entryToJSON(store.entries[index]))
    }

    @MainActor
    private func select(_ body: [String: Any]) async -> [String: Any] {
        let store = NetworkTrafficStore.shared
        if let index = body["index"] as? Int, store.entries.indices.contains(index) {
            store.selectedIndex = index
            return successResponse(store.entryToJSON(store.entries[index]))
        }
        if let pattern = body["urlPattern"] as? String {
            if let idx = store.entries.lastIndex(where: { $0.url.contains(pattern) }) {
                store.selectedIndex = idx
                return successResponse(store.entryToJSON(store.entries[idx]))
            }
            return errorResponse(code: "NOT_FOUND", message: "No request matching '\(pattern)'")
        }
        return errorResponse(code: "MISSING_PARAM", message: "index or urlPattern is required")
    }

    @MainActor
    private func current() async -> [String: Any] {
        let store = NetworkTrafficStore.shared
        guard let entry = store.selectedEntry else {
            return errorResponse(code: "NONE_SELECTED", message: "No request currently selected")
        }
        return successResponse(store.entryToJSON(entry))
    }

    @MainActor
    private func clear() async -> [String: Any] {
        NetworkTrafficStore.shared.clear()
        return successResponse(["cleared": true])
    }
}

File: apps/android/app/src/main/java/com/kelpie/browser/handlers/NetworkInspectorHandler.kt
package com.kelpie.browser.handlers

import com.kelpie.browser.browser.NetworkTrafficStore
import com.kelpie.browser.network.Router
import com.kelpie.browser.network.errorResponse
import com.kelpie.browser.network.successResponse

class NetworkInspectorHandler(
    private val ctx: HandlerContext,
) {
    fun register(router: Router) {
        router.register("network-list") { list(it) }
        router.register("network-detail") { detail(it) }
        router.register("network-select") { select(it) }
        router.register("network-current") { current() }
        router.register("network-clear") { clear() }
    }

    private suspend fun list(body: Map<String, Any?>): Map<String, Any?> {
        val entries =
            NetworkTrafficStore.toSummaryList(
                method = body["method"] as? String,
                category = body["category"] as? String,
                statusRange = body["statusRange"] as? String,
                urlPattern = body["urlPattern"] as? String,
            )
        return successResponse(mapOf("entries" to entries, "total" to NetworkTrafficStore.entries.value.size))
    }

    private suspend fun detail(body: Map<String, Any?>): Map<String, Any?> {
        val index =
            (body["index"] as? Number)?.toInt()
                ?: return errorResponse("INVALID_INDEX", "index is required and must be valid")
        val list = NetworkTrafficStore.entries.value
        if (index !in list.indices) return errorResponse("INVALID_INDEX", "index out of range")
        return successResponse(NetworkTrafficStore.entryToMap(list[index]))
    }

    private suspend fun select(body: Map<String, Any?>): Map<String, Any?> {
        val index = (body["index"] as? Number)?.toInt()
        val list = NetworkTrafficStore.entries.value
        if (index != null && index in list.indices) {
            NetworkTrafficStore.select(index)
            return successResponse(NetworkTrafficStore.entryToMap(list[index]))
        }
        val pattern = body["urlPattern"] as? String
        if (pattern != null) {
            val idx = list.indexOfLast { it.url.contains(pattern) }
            if (idx >= 0) {
                NetworkTrafficStore.select(idx)
                return successResponse(NetworkTrafficStore.entryToMap(list[idx]))
            }
            return errorResponse("NOT_FOUND", "No request matching '$pattern'")
        }
        return errorResponse("MISSING_PARAM", "index or urlPattern is required")
    }

    private suspend fun current(): Map<String, Any?> {
        val entry =
            NetworkTrafficStore.selectedEntry
                ?: return errorResponse("NONE_SELECTED", "No request currently selected")
        return successResponse(NetworkTrafficStore.entryToMap(entry))
    }

    private suspend fun clear(): Map<String, Any?> {
        NetworkTrafficStore.clear()
        return successResponse(mapOf("cleared" to true))
    }
}
